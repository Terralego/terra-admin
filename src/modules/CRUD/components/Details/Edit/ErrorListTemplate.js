import React from 'react';
import PropTypes from 'prop-types';
import { Callout } from '@blueprintjs/core';

const ErrorListTemplate = ({ errorSchema: { __errors, ...fields }, schema: { properties } }) => {
  const errors = Object.keys(fields).reduce((list, item) => {
    const { __errors: errorsItem } = fields[item];
    return errorsItem.length
      ? [...list, { title: properties[item].title, list: errorsItem }]
      : list;
  }, []);

  return (
    <Callout intent="danger">
      <ul>
        {errors.map(({ title, list }) => (
          <li key={title}>
            {title} :
            <ul>
              {list.map(item => <li key={item}>{item}</li>)}
            </ul>
          </li>
        ))}
      </ul>
    </Callout>
  );
};

ErrorListTemplate.propTypes = {
  errorSchema: PropTypes.shape({}).isRequired,
  schema: PropTypes.shape({
    properties: PropTypes.shape({}),
  }).isRequired,
};

export default ErrorListTemplate;
