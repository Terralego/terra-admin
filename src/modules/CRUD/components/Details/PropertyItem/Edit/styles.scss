@import "~@blueprintjs/core/src/components/forms/common";

.CRUD-edit {
  display: flex;
  flex-wrap: wrap;
  justify-content: flex-end;
  align-items: flex-end;
  &--vertical {
    flex-direction: column;
    align-items: stretch;
  }
  &--block > *:first-child {
    flex-basis: 100%;
  }
  fieldset {
    border: 0;
    margin: 0;
    padding: 0;
  }
  .error-list {
    margin: 0.5rem 1.5rem 1rem 0;
  }
  .control-label {
    display: block;
    margin-bottom: 0.2rem;
  }
  .form-group {
    margin-top: 0.4rem;
    margin-right: 0.2rem;
    flex-grow: 1;
    min-width: 0;
    &:first-child {
      margin-top: 0;
    }
    > * {
      width: 100%;
    }
  }
  .checkbox,
  .field-boolean {
    margin-top: 0.3rem;
    input {
      margin-right: 0.3rem;
    }
  }
  .radio-inline {
    margin-right: 1rem;
  }
  .field-radio-group {
    margin: 0.5rem 0;
  }
  input {
    &[type="text"],
    &[type="date"],
    &[type="number"] {
      @include pt-input();
    }
    &:disabled,
    &[readonly] {
      @include pt-input-disabled();
    }
    @include pt-input-placeholder();
  }
  select {
    @include pt-input();
    padding-right: 1rem;
    // Background below is the base 64 transformation of:
    // <svg viewBox="0 0 10 10">
    //   <polygon fill="#5c7080" points="5 0, 10 4, 0 4"/>
    //   <polygon fill="#5c7080" points="5 10, 10 6, 0 6"/>
    // </svg>
    background-image: url(data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgMTAgMTAiIHhtbG5zPSJodHRwOi8vd3d3LnczLm9yZy8yMDAwL3N2ZyI+ICA8cG9seWdvbiBmaWxsPSIjNWM3MDgwIiBwb2ludHM9IjUgMCwgMTAgNCwgMCA0Ii8+ICA8cG9seWdvbiBmaWxsPSIjNWM3MDgwIiBwb2ludHM9IjUgMTAsIDEwIDYsIDAgNiIvPjwvc3ZnPg==);
    background-position: calc(100% - 0.3rem) 50%;
    background-repeat: no-repeat;
    background-size: 0.6rem;
  }
  .help-block {
    opacity: 0.8;
  }
  .error-detail {
    padding: 0.2rem 0 0 1rem;
    color: red;
  }
  .field-array {
    legend {
      font-weight: bold;
      margin-bottom: 0.4rem;
    }
    .array-item,
    .array-item-add {
      display: flex;
      margin: 0.5rem 0 0;
      // Fields
      .col-xs-9 {
        flex-grow: 1;
      }
      // Controls
      .col-xs-3 {
        flex-shrink: 0;
      }
      .btn {
        @include pt-button-base();
        @include pt-button-height($pt-button-height);
        @include pt-button();
        margin-left: 0.2rem;

        &:disabled,
        &.#{$ns}-disabled {
          cursor: not-allowed;
        }
        &::before {
          margin: 0;
          color: #5c7080;
        }
        &.array-item-move-up::before {
          content: "↑";
        }
        &.array-item-move-down::before {
          content: "↓";
        }
        &.array-item-remove::before {
          content: "⨯";
          font-size: 1.5rem;
          margin-top: -0.5rem;
        }
        &.btn-add::before {
          content: "+";
          font-size: 1.1rem;
          margin-top: -0.3rem;
        }
        i {
          display: none;
        }
      }
    }
    .field-array-of-object .array-item-toolbox {
      padding-top: 1.3rem;
    }
  }
  .field-error .form-control {
    border: 1px solid #c23030;
  }
  &__submit {
    flex-shrink: 0;
    .details__list-create & {
      margin: 1rem 0.5rem 0 auto;
    }
  }
}
