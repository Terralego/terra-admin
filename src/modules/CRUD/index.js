import React from 'react';

import config from './config';
import NavLayout from '../../components/NavLayout';
import Nav from './components/Nav';
import Routing from '../../components/Routing';
import CRUDProvider from './services/CRUDProvider';
import MapProvider from './services/MapProvider';
import UserSettingsProvider from './services/UserSettingsProvider';

import './styles.scss';

const { routes } = config;

export const CRUD = () => (
  <UserSettingsProvider>
    <CRUDProvider>
      <MapProvider>
        <div className="CRUD">
          <NavLayout nav={<Nav />}>
            <div className="CRUD-main">
              <Routing
                routes={routes}
              />
            </div>
          </NavLayout>
        </div>
      </MapProvider>
    </CRUDProvider>
  </UserSettingsProvider>
);

CRUD.config = config;

export default CRUD;
