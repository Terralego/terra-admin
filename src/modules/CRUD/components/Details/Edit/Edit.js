import React from 'react';
import PropTypes from 'prop-types';
import Form from 'react-jsonschema-form';
import { Redirect } from 'react-router-dom';
import { Button } from '@blueprintjs/core';
import { CONTROL_DRAW, CONTROLS_TOP_LEFT } from '@terralego/core/modules/Map';
import RTEField from '../../../../../components/react-json-schemaForm/RTEField';

import { ACTION_CREATE, ACTION_UPDATE } from '../../../services/CRUD';
import { toast } from '../../../../../utils/toast';
import { generateURI } from '../../../config';
import Header from '../Header';
import Actions from '../Actions';
import ErrorListTemplate from './ErrorListTemplate';

import {
  ALL,
  POINT,
  LINESTRING,
  POLYGON,
  MULTI_POINT,
  MULTI_LINESTRING,
  MULTI_POLYGON,
} from '../../../../../utils/geom';

const updateSchemaPropertiesValues = (properties, formData) => (
  Object.keys(properties).reduce((list, prop) => ({
    ...list,
    [prop]: {
      ...properties[prop],
      ...(properties[prop].type === 'object')
        ? { properties: updateSchemaPropertiesValues(properties[prop].properties, formData[prop]) }
        : { default: formData[prop] },
    },
  }), {})
);
class Edit extends React.Component {
  static propTypes = {
    map: PropTypes.shape({}),
    feature: PropTypes.shape({
      title: PropTypes.string,
      properties: PropTypes.shape({}),
      geom: PropTypes.shape({}),
    }),
    saveFeature: PropTypes.func.isRequired,
    view: PropTypes.shape({
      formSchema: PropTypes.shape({}),
      layer: PropTypes.shape({ name: PropTypes.string }),
      name: PropTypes.string,
      uiSchema: PropTypes.shape({}),
    }),
    layerPaint: PropTypes.shape({}).isRequired,
    paramLayer: PropTypes.string.isRequired,
    paramId: PropTypes.string.isRequired,
    action: PropTypes.oneOf([ACTION_CREATE, ACTION_UPDATE]).isRequired,
    updateControls: PropTypes.func,
    getSettings: PropTypes.func,
    settingsEndpoint: PropTypes.string,
    displayAddFeature: PropTypes.bool,
    displayChangeFeature: PropTypes.bool,
    history: PropTypes.shape({
      push: PropTypes.func,
    }),
    t: PropTypes.func,
  }

  static defaultProps = {
    map: {},
    feature: {
      title: undefined,
      properties: {},
      geom: {},
    },
    view: {
      formSchema: {},
      layer: { name: undefined },
      name: undefined,
      uiSchema: {},
    },
    displayAddFeature: true,
    displayChangeFeature: true,
    updateControls () {},
    getSettings () {},
    settingsEndpoint: undefined,
    history: {
      push () {},
    },
    t: text => text,
  }

  state = {
    loading: false,
    schema: {
      properties: {},
    },
    geom: {},
  }

  componentDidMount () {
    this.setSchema();

    const {
      action,
      feature: { geom },
    } = this.props;

    if (action !== ACTION_UPDATE || Object.keys(geom).length) {
      this.initDraw();
    }
  }

  componentDidUpdate ({
    paramId: prevParamId,
    action: prevAction,
    feature: prevFeature,
    map: prevMap,
    view: prevView,
    featureError: prevFeatureError,
  }) {
    const {
      action,
      paramId,
      feature,
      map,
      view,
      featureError,
      t,
    } = this.props;
    if (prevFeature !== feature || prevMap !== map) {
      this.initDraw();
    }

    if (
      prevView !== view
      || prevParamId !== paramId
      || prevAction !== action
    ) {
      this.setSchema();
    }

    if (prevFeatureError !== featureError) {
      const { error: { data } } = featureError;
      toast.displayError(
        <div>
          {t(this.isActionUpdate ? 'CRUD.details.failUpdateFeature' : 'CRUD.details.failCreateFeature')}
          {Object.keys(data).map(item => <p key={item}>{item}: {data[item]}</p>)}
        </div>,
      );
    }
  }

  componentWillUnmount () {
    const {
      map,
      updateControls,
    } = this.props;
    // Remove controlDraw from controls
    updateControls([]);
    if (this.layerId && Object.keys(map).length) {
      map.setFilter(this.layerId, ['all']);
    }
  }

  get layerId () {
    const { layerPaint: { id } } = this.props;
    return id;
  }

  buildSchema = (schemaProperties, properties = {}) => {
    const {
      action,
    } = this.props;
    return Object.keys(schemaProperties).reduce((list, prop) => ({
      ...list,
      [prop]: {
        ...schemaProperties[prop],
        ...(schemaProperties[prop].type === 'object')
          ? { properties: this.buildSchema(schemaProperties[prop].properties, properties[prop]) }
          : {},
        ...(action !== ACTION_CREATE && properties[prop] && schemaProperties[prop].type !== 'object')
          ? { default: properties[prop] }
          : {},
      },
    }), {});
  }

  isActionUpdate = () => {
    const { action } = this.props;
    return action === ACTION_UPDATE;
  }

  setSchema = () => {
    const {
      feature: { properties = {}, geom = {} },
      view: { formSchema: schema = {} },
      t,
    } = this.props;
    if (!Object.keys(properties).length && !Object.keys(schema).length) {
      return;
    }
    this.setState(state => {
      const geometryFromMap = {
        type: 'boolean',
        title: t('CRUD.details.geometry'),
        default: !!Object.keys(state.geom).length,
      };
      return {
        geom,
        schema: {
          type: 'object',
          ...schema,
          properties: {
            ...this.buildSchema(schema.properties, properties),
            geometryFromMap,
          },
        },
      };
    });
  }

  initDraw = () => {
    const {
      map,
      paramId,
      feature: { geom },
      updateControls,
      action,
      view: { layer: { geom_type: geomType } },
    } = this.props;

    const control = {
      control: CONTROL_DRAW,
      position: CONTROLS_TOP_LEFT,
      onDrawUpdate: this.updateGeometry,
      onDrawCreate: this.updateGeometry,
      onDrawDelete: this.updateGeometry,
      controls: {
        point:  [ALL, POINT, MULTI_POINT].includes(geomType),
        line_string: [ALL, LINESTRING, MULTI_LINESTRING].includes(geomType),
        polygon:  [ALL, POLYGON, MULTI_POLYGON].includes(geomType),
        trash: true,
        combine_features: false,
        uncombine_features: false,
      },
    };

    if (action === ACTION_UPDATE) {
      if (!geom || !Object.keys(map).length) return;

      const listener = ({ control: addedControl }) => {
        if (addedControl !== control.control) {
          return;
        }
        map.draw.add(geom);
        map.off('control_added', listener);
      };
      map.on('control_added', listener);
      if (this.layerId) {
        map.setFilter(this.layerId, ['!=', '_id', paramId]);
      }
    }
    updateControls([control]);
  }

  updateGeometry = ({ type, features: [{ geometry: geom, id }], target: map }) => {
    if (type === 'draw.delete') {
      this.setState({ geom: {} });
      return;
    }

    const { view: { layer: { geom_type: geomType } } } = this.props;
    const { features } = map.draw.getAll();

    if (features.length > 1 && [POINT, LINESTRING, POLYGON].includes(geomType)) {
      map.draw.delete(
        features.reduce((list, feature) => (
          feature.id !== id
            ? [...list, feature.id]
            : list
        ), []),
      );
    }

    this.setState({ geom });
  }

  changeForm = ({ formData }) => {
    const { t } = this.props;

    this.setState(({ schema, geom }) => ({
      schema: {
        ...schema,
        properties: {
          ...updateSchemaPropertiesValues(schema.properties, formData),
          geometryFromMap: {
            type: 'boolean',
            title: t('CRUD.details.geometry'),
            default: !!Object.keys(geom).length,
          },
        },
      },
    }));
  }

  submitFeature = async ({ formData: { geometryFromMap, ...properties } }) => {
    const { geom } = this.state;
    const {
      getSettings,
      history: { push },
      view: { featureEndpoint },
      paramId,
      paramLayer,
      getFeaturesList,
      saveFeature,
      settingsEndpoint,
      t,
    } = this.props;

    this.setState({
      loading: true,
    });

    const savedFeature = await saveFeature(
      featureEndpoint,
      this.isActionUpdate ? paramId : false,
      { geom, properties },
    );

    if (savedFeature) {
      push(generateURI('layer', { layer: paramLayer, id: savedFeature.identifier }));
      getFeaturesList(
        featureEndpoint,
        { querystring: { page_size: 2000 } },
      );
      toast.displayToaster(
        { id: savedFeature.identifier },
        t(this.isActionUpdate ? 'CRUD.details.successUpdateFeature' : 'CRUD.details.successCreateFeature'),
      );
    } else {
      this.setState({
        loading: false,
      });
    }


    getSettings(settingsEndpoint);
  }

  validateForm = (formData, errors) => {
    const { geom } = this.state;
    const { t } = this.props;
    if (!Object.keys(geom).length) {
      errors.geometryFromMap.addError(t('CRUD.details.errorNoGeometry'));
    }
    return errors;
  }

  render () {
    const { loading, schema, schema: { properties } } = this.state;
    const {
      t,
      action,
      view: { layer: { name }, name: displayName = name, uiSchema = {} },
      paramLayer,
      paramId,
      displayAddFeature,
      displayChangeFeature,
      feature: { title },
    } = this.props;

    if (
      (action === ACTION_CREATE && !displayAddFeature)
      || (action === ACTION_UPDATE && !displayChangeFeature)
    ) {
      toast.displayError(t('CRUD.details.noAccess'));
      return (<Redirect to={generateURI('layer', { layer: paramLayer })} />);
    }

    const mainTitle = action === ACTION_CREATE
      ? t('CRUD.details.create', { layer: displayName })
      : (title || t('CRUD.details.noFeature'));

    const SaveButton = props => (
      <Button
        intent="primary"
        className="details__actions-save"
        loading={loading}
        {...props}
      >
        {action === ACTION_CREATE ? mainTitle : t('CRUD.details.save')}
      </Button>
    );

    const actionsButtons = {
      paramLayer,
      displayCancel: true,
      displayDelete: action === ACTION_UPDATE,
      ...action === ACTION_UPDATE ? { paramId } : {},
    };

    return (
      <div className="details ">
        <Header title={mainTitle} />
        <div className="details__content bootstrap-inside">
          {Object.keys(properties).length
            ? (
              <Form
                schema={schema}
                uiSchema={{ ...uiSchema, geometryFromMap: { 'ui:widget': 'hidden' } }}
                fields={{
                  rte: RTEField,
                }}
                onSubmit={this.submitFeature}
                onChange={this.changeForm}
                validate={this.validateForm}
                ErrorList={ErrorListTemplate}
              >
                <Actions {...actionsButtons}>
                  <SaveButton type="submit" />
                </Actions>
              </Form>
            )
            : (
              <Actions {...actionsButtons}>
                <SaveButton onClick={() => this.submitFeature({})} />
              </Actions>
            )
          }
        </div>
      </div>
    );
  }
}

export default Edit;
