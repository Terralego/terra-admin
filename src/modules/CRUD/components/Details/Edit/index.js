import { withRouter } from 'react-router-dom';
import { withTranslation } from 'react-i18next';
import { connectAuthProvider } from '@terralego/core/modules/Auth';

import { connectAppProvider } from '../../../../../components/AppProvider';
import { connectCRUDProvider } from '../../../services/CRUDProvider';
import { connectMapProvider } from '../../../views/Map/MapProvider';
import { withTableFilters } from '../../../services/UserSettingsProvider';
import { getView, getLayersPaints } from '../../../services/CRUD';
import compose from '../../../../../utils/compose';

import Edit from './Edit';

const appProviderGetter = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

const authProviderGetter = ({
  authenticated,
  user,
}) => {
  const permissions = authenticated ? user.permissions : [];
  return {
    displayAddFeature: permissions.includes('can_add_feature'),
    displayChangeFeature: permissions.includes('can_change_feature'),
  };
};

const CRUDPRoviderGetter = ({
  getSettings,
  getFeaturesList,
  settings,
  map,
  feature,
  saveFeature,
  errors: {
    feature: featureError,
  },
}, {
  match: { params: { layer, id } },
}) => ({
  getSettings,
  getFeaturesList,
  map,
  feature: feature[id] || {},
  saveFeature,
  view: getView(settings, layer),
  layerPaint: getLayersPaints(settings).find(item => item['source-layer'] === layer) || {},
  paramLayer: layer,
  paramId: id,
  featureError: featureError.find(({ featureId }) => featureId === id),
});

export default compose(
  withRouter,
  connectAppProvider(appProviderGetter),
  connectAuthProvider(authProviderGetter),
  connectCRUDProvider(CRUDPRoviderGetter),
  connectMapProvider('addControl', 'removeControl'),
  withTableFilters(),
  withTranslation(),
)(Edit);
