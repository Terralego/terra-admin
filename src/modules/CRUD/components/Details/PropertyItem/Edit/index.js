import { connectAppProvider } from '../../../../../../components/AppProvider';

import Edit from './Edit';

import compose from '../../../../../../utils/compose';

const appProviderGetter = ({
  env: { modules: { CRUD: { settings } } },
}) => ({
  settingsEndpoint: settings,
});

export default compose(
  connectAppProvider(appProviderGetter),
)(Edit);
