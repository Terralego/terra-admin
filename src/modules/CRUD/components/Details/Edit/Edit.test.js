import React from 'react';
import renderer from 'react-test-renderer';
import { CONTROL_DRAW } from '@terralego/core/modules/Map';

import { ACTION_CREATE, ACTION_UPDATE } from '../../../services/CRUD';

import Edit from './Edit';

jest.mock('react-jsonschema-form', () => ({ children }) => <form>{children}</form>);

jest.mock('react-router-dom', () => ({
  Redirect: () => <div>Error because Redirect</div>,
}));

jest.mock('@blueprintjs/core', () => ({
  Button: () => <div>Button</div>,
}));

jest.mock('@terralego/core/modules/Map', () => ({
  CONTROL_DRAW: 'DrawControl',
  CONTROLS_TOP_LEFT: 'top-left',
}));

jest.mock('../../../views/Map/Map', () => ({
  ACTION_CREATE: 'create',
  ACTION_UPDATE: 'update',
}));

jest.mock('../../../../../utils/geom', () => ({
  ALL: null,
  POINT: 0,
  LINESTRING: 1,
  POLYGON: 3,
  MULTI_POINT: 4,
  MULTI_LINESTRING: 5,
  MULTI_POLYGON: 6,
}));

jest.mock('../../../../../utils/toast', () => ({
  toast: {
    displayToaster: bool => <div>{bool ? 'Success' : 'Fail'}</div>,
    displayError: children => <div>{children}</div>,
  },
}));

jest.mock('../../../config', () => ({
  generateURI: jest.fn(),
}));

jest.mock('../Actions', () => () => <div>Actions buttons</div>);
jest.mock('./ErrorListTemplate', () => () => <div>Error list template</div>);

let props;
beforeEach(() => {
  props = {
    settings: {},
    map: {},
    feature: {
      title: 'Title of the feature',
      properties: {},
      geom: {},
    },
    saveFeature: jest.fn(),
    updateControls: jest.fn(),
    getFeaturesList: jest.fn(),
    getSettings: jest.fn(),
    view: {
      name: 'foo',
      layer: {
        name: 'layerFoo',
        id: 8,
        geom_type: 0,
      },
      formSchema: {
        properties: {
          city: { type: 'boolean', title: 'City' },
          name: { type: 'text', default: 'Title' },
          group: {
            type: 'object',
            properties: {
              name: { type: 'string', title: 'Group name' },
            },
          },
        },
      },
    },
    layerPaint: { id: 'fooComponentID' },
    paramLayer: 'fooLayer',
    paramId: 'fooId',
    t: key => key,
    action: ACTION_UPDATE,
    displayAddFeature: true,
    displayChangeFeature: true,
    history: {
      push: jest.fn(),
    },
  };
});

describe('Snapshots', () => {
  it('should render correctly', () => {
    const tree = renderer.create((
      <Edit
        {...props}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly without schema props', () => {
    const tree = renderer.create((
      <Edit
        {...props}
        view={{
          ...props.view,
          formSchema: {},
        }}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should not render without permissions', () => {
    const tree = renderer.create((
      <Edit
        {...props}
        layer={{
          name: 'bar',
          id: 8,
          geom_type: 0,
        }}
        displayChangeFeature={false}
      />
    )).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

it('should not init draw', () => {
  const instance = new Edit({
    ...props,
  });
  instance.initDraw = jest.fn();
  instance.componentDidMount();
  expect(instance.initDraw).not.toHaveBeenCalled();
});

it('should init draw', () => {
  const instance = new Edit({
    ...props,
    feature: {
      ...props.feature,
      geom: {
        type: 'Point',
        coordinates: [
          0.619166,
          44.206944,
        ],
      },
    },
    action: ACTION_CREATE,
  });
  instance.initDraw = jest.fn();
  instance.setState = jest.fn();
  instance.componentDidMount();
  expect(instance.initDraw).toHaveBeenCalled();
});

it('should init draw when the feature or map props are changing', () => {
  const instance = new Edit({
    ...props,
    feature: {
      ...props.feature,
      geom: {
        type: 'Point',
        coordinates: [
          0.619166,
          44.206944,
        ],
      },
    },
    action: ACTION_CREATE,
  });
  instance.initDraw = jest.fn();
  instance.componentDidUpdate({
    ...props,
    feature: {
      geom: {
        type: 'LineString',
        coordinates: [
          0.8,
          44.1,
        ],
      },
    },
  });
  expect(instance.initDraw).toHaveBeenCalled();
});

it('should execute init draw', () => {
  const newProps = {
    ...props,
    feature: {
      geom: {
        type: 'Point',
        coordinates: [
          0.619166,
          44.206944,
        ],
      },
    },
    action: ACTION_CREATE,
  };
  const instance = new Edit({ ...newProps });
  instance.initDraw();
  expect(newProps.updateControls).toHaveBeenCalledWith([{
    control: 'DrawControl',
    controls: {
      combine_features: false,
      line_string: false,
      point: true,
      polygon: false,
      trash: true,
      uncombine_features: false,
    },
    onDrawCreate: instance.updateGeometry,
    onDrawDelete: instance.updateGeometry,
    onDrawUpdate: instance.updateGeometry,
    position: 'top-left',
  }]);
});

it('should execute init draw and define map draw', () => {
  let onControlAdded;
  const newProps = {
    ...props,
    feature: {
      geom: {
        type: 'Point',
        coordinates: [
          0.619166,
          44.206944,
        ],
      },
    },
    action: ACTION_UPDATE,
    map: {
      draw: {
        add: jest.fn(),
      },
      on: (_, fn) => {
        onControlAdded = fn;
      },
      off: jest.fn(),
      setFilter: jest.fn(),
    },
  };
  const instance = new Edit({ ...newProps });
  instance.setMapFilter = jest.fn();
  instance.initDraw();
  onControlAdded({ control: CONTROL_DRAW });
  expect(instance.props.map.draw.add).toHaveBeenCalled();
  onControlAdded({ control: 'plop' });
  expect(newProps.updateControls).toHaveBeenCalledWith([{
    control: 'DrawControl',
    controls: {
      combine_features: false,
      line_string: false,
      point: true,
      polygon: false,
      trash: true,
      uncombine_features: false,
    },
    onDrawCreate: instance.updateGeometry,
    onDrawDelete: instance.updateGeometry,
    onDrawUpdate: instance.updateGeometry,
    position: 'top-left',
  }]);
});


it('should update schema when the schema prop is changing', () => {
  const instance = new Edit({
    ...props,
  });
  instance.setSchema = jest.fn();
  instance.componentDidUpdate({
    ...props,
    view: {
      ...props.views,
      formSchema: {
        properties: {
          city: { type: 'boolean', title: 'City' },
          name: { type: 'text', default: 'Title' },
        },
      },
    },
  });
  expect(instance.setSchema).toHaveBeenCalled();
});

it('should update schema', () => {
  const instance = new Edit({
    ...props,
  });
  let stateCallback;
  instance.setState = jest.fn(callback => {
    stateCallback = callback;
  });
  instance.setSchema(props.schema);
  expect(stateCallback(instance.state)).toEqual({
    geom: {},
    schema: {
      properties: {
        city: { title: 'City', type: 'boolean' },
        geometryFromMap: { default: false, title: 'CRUD.details.geometry', type: 'boolean' },
        name: { default: 'Title', type: 'text' },
        group: {
          type: 'object',
          properties: {
            name: { type: 'string', title: 'Group name' },
          },
        },
      },
      type: 'object',
    },
  });
});

it('should update geometry and clear the draw if necessary', () => {
  const instance = new Edit({
    ...props,
  });
  const geom = {
    type: 'draw.update',
    features: [{
      id: 'eae43f270872e8249df87499e079787d',
      type: 'Feature',
      properties: {},
      geometry: {
        coordinates: [
          1.0602676121518189,
          47.36944181389896,
        ],
        type: 'Point',
      },
    }],
    target: {
      draw: {
        getAll: () => ({ features: [{ id: 'foo' }, { id: 'bar' }] }),
        delete: jest.fn(),
      },
    },
  };

  instance.setState = jest.fn();
  instance.updateGeometry(geom);

  expect(instance.setState).toHaveBeenCalledWith({
    geom: { coordinates: [1.0602676121518189, 47.36944181389896], type: 'Point' },
  });
  expect(geom.target.draw.delete).toHaveBeenCalledWith(['foo', 'bar']);
});

it('should update geometry', () => {
  const instance = new Edit({
    ...props,
  });

  const geom = {
    type: 'draw.update',
    features: [{
      id: 'eae43f270872e8249df87499e079787d',
      type: 'Feature',
      properties: {},
      geometry: {
        coordinates: [
          1.0602676121518189,
          47.36944181389896,
        ],
        type: 'Point',
      },
    }],
    target: {
      draw: {
        getAll: () => ({ features: [{ id: 'foo' }] }),
        delete: jest.fn(),
      },
    },
  };

  instance.setState = jest.fn();
  instance.updateGeometry(geom);

  expect(geom.target.draw.delete).not.toHaveBeenCalledWith();
});

it('should delete geometry', () => {
  const instance = new Edit({
    ...props,
  });

  const geom = {
    type: 'draw.delete',
    features: [{
      id: 'eae43f270872e8249df87499e079787d',
      type: 'Feature',
      properties: {},
      geometry: {
        coordinates: [
          1.0602676121518189,
          47.36944181389896,
        ],
        type: 'Point',
      },
    }],
    target: {
      draw: {
        getAll: () => ({ features: [{ id: 'foo' }] }),
        delete: jest.fn(),
      },
    },
  };

  instance.setState = jest.fn();
  instance.updateGeometry(geom);

  expect(instance.setState).toHaveBeenCalledWith({ geom: {} });
});

it('should change form', () => {
  const instance = new Edit({
    ...props,
  });
  instance.state = {
    geom: { coordinates: [1.0602676121518189, 47.36944181389896], type: 'Point' },
    schema: {
      properties: {
        city: {
          type: 'string',
          title: 'Ville',
        },
        name: {
          type: 'string',
          title: 'Nom',
        },
      },
    },
  };

  let stateCallback;
  instance.setState = jest.fn(callback => {
    stateCallback = callback;
  });
  instance.setSchema(props.view.formSchema);
  instance.changeForm({ formData: { city: 'Toulouse', geometryFromMap: false } });

  expect(stateCallback(instance.state)).toEqual({
    schema: {
      properties: {
        geometryFromMap: { default: true, title: 'CRUD.details.geometry', type: 'boolean' },
        city: { default: 'Toulouse', title: 'Ville', type: 'string' },
        name: { default: undefined, title: 'Nom', type: 'string' },
      },
    },
  });
});

it('should submit form', async () => {
  const instance = new Edit({
    ...props,
    action: ACTION_UPDATE,
    saveFeature: jest.fn(() => ({ id: 'Saved!' })),
  });
  instance.state = {
    geom: { coordinates: [1.0602676121518189, 47.36944181389896], type: 'Point' },
  };
  instance.setState = jest.fn();
  instance.submitFeature({ formData: { city: 'Toulouse', geometryFromMap: false } });
  await true;
  expect(instance.setState).toHaveBeenCalledTimes(1);
  expect(instance.props.history.push).toHaveBeenCalled();
});

it('should not submit form', async () => {
  const instance = new Edit({
    ...props,
    action: ACTION_UPDATE,
    saveFeature: jest.fn(() => undefined),
  });
  instance.state = {
    geom: { coordinates: [1.0602676121518189, 47.36944181389896], type: 'Point' },
  };
  instance.setState = jest.fn();
  instance.submitFeature({ formData: { city: 'Toulouse', geometryFromMap: false } });
  await true;
  expect(instance.setState).toHaveBeenCalledTimes(2);
  expect(instance.props.history.push).not.toHaveBeenCalled();
});

it('should validate form', () => {
  const instance = new Edit({
    ...props,
  });
  instance.state = {
    geom: { coordinates: [1.0602676121518189, 47.36944181389896], type: 'Point' },
  };
  expect(instance.validateForm({}, {})).toEqual({});
});

it('should not validate form', () => {
  const instance = new Edit({
    ...props,
  });
  instance.state = {
    geom: {},
  };
  const errors = {
    geometryFromMap: {
      addError: jest.fn(),
    },
  };
  expect(instance.validateForm({}, errors)).toEqual(errors);
});

it('should remove control when component will unmout', () => {
  const instance = new Edit({
    ...props,
  });
  instance.setMapFilter = jest.fn();
  instance.componentWillUnmount();
  expect(instance.props.updateControls).toHaveBeenCalled();
  expect(instance.setMapFilter).toHaveBeenCalledWith('fooComponentID', ['all']);
});
