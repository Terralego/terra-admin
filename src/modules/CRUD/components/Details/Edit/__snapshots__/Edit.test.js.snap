// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshots should not render without permissions 1`] = `
<div>
  Error because Redirect
</div>
`;

exports[`Snapshots should render correctly 1`] = `
<div
  className="details "
>
  <div
    className="details__header"
  >
    <h2
      className="details__title"
    >
      Title
    </h2>
  </div>
  <div
    className="details__content bootstrap-inside"
  >
    <form>
      <div>
        Actions buttons
      </div>
    </form>
  </div>
</div>
`;

exports[`Snapshots should render correctly without schema props 1`] = `
<div
  className="details "
>
  <div
    className="details__header"
  >
    <h2
      className="details__title"
    >
      CRUD.details.noFeature
    </h2>
  </div>
  <div
    className="details__content bootstrap-inside"
  >
    <div>
      Actions buttons
    </div>
  </div>
</div>
`;
